### PR TITLE
HttpClient Component [wip]

### PR DESCRIPTION
**Pull request by SuddenlyHazel**

### CODE REVIEW

File: guests/function/src/lib.rs [hunk 1]
```diff
@@ -12,7 +14,9 @@ wit_bindgen::generate!({
     // namely the `run` function.
     exports: {
         world: TestFunction,
-        "mycelia:execution/function-interface": TestFunction
+        "mycelia:execution/function-world": TestFunction,
```
Comment (SuddenlyHazel) [inline guests/function/src/lib.rs:17]: TODO remove these. Was experimenting..

File: guests/function/src/lib.rs [hunk 2]
```diff
@@ -26,8 +30,23 @@ wit_bindgen::generate!({
 // is passed to it. Or, returns "hello world"
 struct TestFunction;
 
+pub struct MyTestHandler;
```
Comment (SuddenlyHazel) [inline guests/function/src/lib.rs:33]: TODO remove these

File: services/function/src/lib.rs [hunk 3]
```diff
@@ -310,7 +319,24 @@ mod test {
...
+
+            Self {
+                table,
+                ctx,
+                http_client_maker,
+            }
+        }
+    }
+
+    impl HostClientResourceMaker for ServerWasiView {
```
Comment (SuddenlyHazel) [inline services/function/src/lib.rs:337]: After all this work - We now have a pattern that lets us inject wasm component resources to be used by guests and all we have to do runtime side is

implement a trait 

and..

File: services/function/src/lib.rs [hunk 4]
```diff
@@ -327,6 +353,13 @@ mod test {
 
         let _ = add_to_linker(&mut linker)?;
 
+        let _ = resource_providers::http::setup_with_wasmtime(
```
Comment (SuddenlyHazel) [inline services/function/src/lib.rs:356]: Connect it to the component, store, and linker :D

File: guest_crates/mycelia_http/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+license = "MIT OR Apache-2.0"
+description = "guest side wasm http-client for general use in a compatible wasm component host. See docs for more info!"
+authors = ["Hazel Rowell <hazel+mycelia@departure.dev>"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+thiserror = { workspace = true }
+async-trait = { workspace = true }
+anyhow = { workspace = true }
```
Comment (danielterwiel) [inline guest_crates/mycelia_http/Cargo.toml:14]: How do you decide which crate to use for which type of Errors?

I've been reading [this article](https://www.shakacode.com/blog/thiserror-anyhow-or-how-i-handle-errors-in-rust-apps/#:~:text=Use%20thiserror%20%2C%20when%20callers%20are,it%20to%20the%20logging%20pipeline.) and I quote:

> When to use which? That's how I've been making my decisions:

> Use thiserror, when callers are interested in error details. E.g., if an error is an enum, a caller has to handle each error variant differently.
> Use anyhow, when the caller doesn't care about the error details. E.g., it just propagates it to the logging pipeline.
> I've been using both libraries, and my experience with anyhow turned out not to be great. While it is convenient operating on a single type, error handling in apps becomes messy over time, which makes it hard to hold errors quality under control across the app.

File: guest_crates/mycelia_http/wit/deps.toml [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (danielterwiel) [inline guest_crates/mycelia_http/wit/deps.toml:1]: nit: can we keep dependency declarations in the cargo.toml and rm this file + its lockifile?

File: resource_providers/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+thiserror = { workspace = true }
+async-trait = { workspace = true }
+anyhow = { workspace = true }
+wit-bindgen = { workspace = true }
+http = { workspace = true }
+wasmtime = { workspace = true }
+wasmtime-wasi = { workspace = true }
+tower = { version = "0.4.13", features = ["util"] }
+hyper = { version = "0.14.27", features = ["full"]}
+tokio = { workspace = true, features = ["full"]}
```
Comment (danielterwiel) [inline resource_providers/Cargo.toml:18]: What's the reasoning behind using `workspace = true` in some cases and fixed versioning/featuring in other cases. Tower and Hyper can also be shared workspace-wide, right?

File: resource_providers/src/http.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /// Attempts to make an HttpRequest using some resource
+    async fn send(
+        &mut self,
+        guest_self: Resource<Client>,
+        req: ClientRequest,
+    ) -> anyhow::Result<ClientResult> {
+        let id = guest_self.rep();
+        let client = self.clients.get_mut(&id).ok_or(HttpClientError::HostResourceNotFound)?;
```
Comment (danielterwiel) [inline resource_providers/src/http.rs:139]: Why aren't we `panic`'ing on `HostResourceNotFound`? Its comment states that it should never happen

File: resource_providers/src/http.rs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+            clients: Default::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl HostClientInterface for HostClientResource {
+    /// Creates a new HttpClient Resource and returns the resource to the guest.
+    ///
+    /// We return a resource id the guest will use to request access to the resource
```
Comment (danielterwiel) [inline resource_providers/src/http.rs:111]:
```suggestion
    /// We return a resource id which the guest will use to request access to the resource
```

File: resource_providers/src/http.rs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    /// internally, we create a mapping of <ID, Client> so we can lookup the correct
+    /// client to make the request.
```
Comment (danielterwiel) [inline resource_providers/src/http.rs:113]: This ID appears to live in the session, is that correct? Should it be backed up for data integrity reasons?

File: xtask/src/main.rs [hunk 11]
```diff
@@ -177,6 +177,7 @@ fn build_component(guest: &Guest) -> Result<(), DynError> {
             &cmd_wasm_guest,
             &cmd_wasi_snapshot,
             &cmd_component_output,
+            "-vvv"
```
Comment (danielterwiel) [inline xtask/src/main.rs:180]: What does this arg do? Cannot find it very easily

File: guest_crates/mycelia_http/wit/deps.toml (hunk 6 above)
Comment (SuddenlyHazel) [reply guest_crates/mycelia_http/wit/deps.toml:1]: This is needed for `wit-deps` which relies on this known structure still :\

File: resource_providers/Cargo.toml (hunk 7 above)
Comment (SuddenlyHazel) [reply resource_providers/Cargo.toml:18]: Good call, I can lock those in the workspace.

File: resource_providers/src/http.rs (hunk 8 above)
Comment (SuddenlyHazel) [reply resource_providers/src/http.rs:139]: I'm not entirely sure how we should handle this case tbh - lets panic. Good call

File: resource_providers/src/http.rs (hunk 10 above)
Comment (SuddenlyHazel) [reply resource_providers/src/http.rs:113]: These resource IDs are ephemeral

File: xtask/src/main.rs (hunk 11 above)
Comment (SuddenlyHazel) [reply xtask/src/main.rs:180]: ahh actually need to remove this. This just enables trace logs

File: guest_crates/mycelia_http/Cargo.toml (hunk 5 above)
Comment (SuddenlyHazel) [reply guest_crates/mycelia_http/Cargo.toml:14]: I've really been enjoying thiserror. I'm not sure I'm confident enough quite yet to make a total switch tho

**Pull request merged**